File: nodejs/utils/random_name.mjs
import { randomBytes } from 'node:crypto';

const LOWER = 'abcdefghijklmnopqrstuvwxyz';
const UPPER = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
const DIGITS = '0123456789';

/**
 * Generate a random string.
 * @param {number} length positive integer length
 * @param {{lettersOnly?: boolean, uppercase?: boolean}} [options]
 * @returns {string}
 */
export function randomName(length, { lettersOnly = false, uppercase = false } = {}) {
  if (!Number.isInteger(length) || length <= 0) {
    throw new TypeError('length must be a positive integer');
  }
  if (!lettersOnly) {
    // hex of random bytes, then slice
    return randomBytes(Math.ceil(length / 2)).toString('hex').slice(0, length);
  }

  const alphabet = (uppercase ? UPPER : LOWER);
  const bytes = randomBytes(length);
  let out = '';
  for (let i = 0; i < length; i++) {
    out += alphabet[bytes[i] % alphabet.length];
  }
  return out;
}

// CLI usage: node random_name.mjs <length> [lettersOnly] [uppercase]
if (import.meta.url === `file://${process.argv[1]}`) {
  const length = Number(process.argv[2]);
  const lettersOnly = process.argv[3] ? process.argv[3] === 'true' : false;
  const uppercase = process.argv[4] ? process.argv[4] === 'true' : false;
  try {
    console.log(randomName(length, { lettersOnly, uppercase }));
  } catch (err) {
    console.error(err.message);
    process.exit(1);
  }
}

File: go/utils/random_name.go
package main

import (
    "crypto/rand"
    "encoding/hex"
    "fmt"
    "math/big"
    "os"
    "strconv"
)

func randomName(length int, lettersOnly bool, uppercase bool) (string, error) {
    if length <= 0 {
        return "", fmt.Errorf("length must be positive")
    }
    if !lettersOnly {
        buf := make([]byte, (length+1)/2)
        if _, err := rand.Read(buf); err != nil {
            return "", err
        }
        hexStr := hex.EncodeToString(buf)
        return hexStr[:length], nil
    }
    lower := "abcdefghijklmnopqrstuvwxyz"
    upper := "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
    alphabet := lower
    if uppercase {
        alphabet = upper
    }
    out := make([]byte, length)
    for i := 0; i < length; i++ {
        n, err := rand.Int(rand.Reader, big.NewInt(int64(len(alphabet))))
        if err != nil {
            return "", err
        }
        out[i] = alphabet[n.Int64()]
    }
    return string(out), nil
}

func main() {
    if len(os.Args) < 2 {
        fmt.Println("usage: random_name <length> [lettersOnly] [uppercase]")
        os.Exit(1)
    }
    length, _ := strconv.Atoi(os.Args[1])
    lettersOnly := len(os.Args) > 2 && os.Args[2] == "true"
    uppercase := len(os.Args) > 3 && os.Args[3] == "true"
    s, err := randomName(length, lettersOnly, uppercase)
    if err != nil {
        fmt.Println(err)
        os.Exit(1)
    }
    fmt.Println(s)
}


